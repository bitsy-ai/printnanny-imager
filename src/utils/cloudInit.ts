import * as YAML from 'yaml';
import bcrypt from "bcrypt";

interface CloudInitForm {
  disableSSHPassword: boolean;
  enableRemoteVideo: boolean;
  enableSSH: boolean;
  enableTelemetry: boolean;
  hostname: string;
  keyboardLayout: string;
  password: string;
  saveSettings: boolean;
  timezone: string;
  username: string;
  wifiCountry: string;
  wifiPassword: string;
  wifiSSID: string;
  wifiSSIDHidden: boolean;
  sshAuthorizedKeys?: string;
}

interface PasswordHashOptions {
    algorithm: string;
    saltLength: string;
    iterations: number;
}

class CloudInitGenerator implements CloudInitForm {
  disableSSHPassword: boolean;
  enableRemoteVideo: boolean;
  enableSSH: boolean;
  enableTelemetry: boolean;
  hostname: string;
  keyboardLayout: string;
  password: string;
  saveSettings: boolean;
  timezone: string;
  username: string;
  wifiCountry: string;
  wifiPassword: string;
  wifiSSID: string;
  wifiSSIDHidden: boolean;
  sshAuthorizedKeys?: string;

  constructor(args: CloudInitForm) {
    this.sshAuthorizedKeys = args.sshAuthorizedKeys;
    this.disableSSHPassword = args.disableSSHPassword;
    this.enableRemoteVideo = args.enableRemoteVideo;
    this.enableSSH = args.enableSSH;
    this.enableTelemetry = args.enableTelemetry;
    this.hostname = args.hostname;
    this.keyboardLayout = args.keyboardLayout;
    this.password = args.password;
    this.saveSettings = args.saveSettings;
    this.timezone = args.timezone;
    this.username = args.username;
    this.wifiCountry = args.wifiCountry;
    this.wifiPassword = args.wifiPassword;
    this.wifiSSID = args.wifiSSID;
    this.wifiSSIDHidden = args.wifiSSIDHidden;
  }

  static comparePassword(password: string, hash: string): boolean {
    return bcrypt.compareSync(password, hash);
  }

  static hashPassword(password: string, saltRounds = 4096): string {
    return bcrypt.hashSync(password, saltRounds);
  }

  static defaultGroups() {
    return [
      "adm",
      "audio",
      "cdrom",
      "games",
      "gpio",
      "i2c",
      "input",
      "netdev",
      "plugdev",
      "render",
      "spi",
      "sudo",
      "users",
      "video",
    ];
  }

  static defaultShell() {
    return "/bin/bash";
  }

  generateUserBlock() {
    const userBlock = {
      name: this.username,
      groups: CloudInitGenerator.defaultGroups(),
      lock_passwd: false,
      passwd: this.password, // password is already hashed before class is constructed
      shell: CloudInitGenerator.defaultShell(),
      sudo: 'ALL=(ALL) NOPASSWD:ALL',
      ssh_authorized_keys: [] as string[]
    };
    if (this.sshAuthorizedKeys !== undefined){
        userBlock.ssh_authorized_keys=  this.sshAuthorizedKeys.split('\n');
    }
    return userBlock
  }

  generateUserData() {
    // create json data structure
    const jsonData = { hostname: this.hostname, manage_etc_hosts: true, users: [ "default", this.generateUserBlock()], timezone: this.timezone };
    // convert to yaml
    const yamlData = YAML.stringify(jsonData);
    // add #cloud-config to first line of generated yaml
    return `#cloud-config
${yamlData}
`
  }
}

export { CloudInitGenerator };
export type { CloudInitForm };
