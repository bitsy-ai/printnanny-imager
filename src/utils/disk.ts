import { Command } from '@tauri-apps/api/shell'
import { platform } from '@tauri-apps/api/os';
import { RemoveableDisk, RemoveableDiskPartition } from '../types';

function parseLinuxDisks(jsonStr: string): Array<RemoveableDisk> {
  const parsed = JSON.parse(jsonStr);
  console.log(parsed)
  if (parsed.blockdevices == undefined) { console.error("Failed to parse lsblk output", parsed); return [] }
  const usbDevices = parsed.blockdevices.filter((device: any) => device.tran == "usb");
  if (usbDevices.length == 0) { return [] }
  return usbDevices.map((device: any) => {
    const partitions = device.children && device.children.length > 0 ? device.children.map((part: any) => {
      return {
        name: part.name,
        label: part.label,
        size: part.size,
        mountpoint: part.mountpoint,
      } as RemoveableDiskPartition
    }) : [];

    return {
      key: device.name,
      name: device.name,
      serial: device.serial,
      model: device.model,
      vendor: device.vendor,
      partitions: partitions,
      size: device.size
    } as RemoveableDisk
  })
}

async function listRemoveableDisks(): Promise<Array<RemoveableDisk>> {
  const platformName = await platform();
  switch (platformName) {
    case 'darwin':
      const macOSOutput = await new Command('list-diskdrive--macOS').execute();
      console.log(JSON.parse(macOSOutput.stdout));
      return []
      break
    case 'linux':
      const linuxOutput = await new Command('list-diskdrive--linux').execute();
      return parseLinuxDisks(linuxOutput.stdout)
    case 'win32':
      const output = await new Command('list-diskdrive--windows').execute();
      console.log(JSON.parse(output.stdout));
      return []
      break
    default:
      console.error(`${platformName} is not supported. Please open a Github issue to request support.`)
      return []
  }
}

export {
  listRemoveableDisks,
  parseLinuxDisks
}