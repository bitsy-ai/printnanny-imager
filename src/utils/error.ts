import { useRouter } from "vue-router";
import { useStore } from "@/store";
import { AlertAction, UiAlert } from "@/types";
import { ExclamationTriangleIcon } from "@heroicons/vue/24/outline";

const ResetAction = {
  color: "indigo",
  text: "Reset",
  onClick: () => {
    const store = useStore();
    const router = useRouter();
    store.$reset();
    router.push({ name: "select-image" });
  },
} as AlertAction;

function showError(e: Error) {
  console.error(e);
  // handle actione rror
  const store = useStore();
  const message = e.toString();
  const header = "Error";

  const alert = {
    message,
    header,
    icon: ExclamationTriangleIcon,
    actions: [ResetAction],
  } as UiAlert;

  store.showError(alert);
}

export { showError, ResetAction };
