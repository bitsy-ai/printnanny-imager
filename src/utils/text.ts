function truncate(str: string): string {
    return str.slice(0, 10) + "..." + str.slice(str.length - 10, str.length);
}

export { truncate }