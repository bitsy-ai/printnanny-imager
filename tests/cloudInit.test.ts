import { describe, expect, test } from "@jest/globals";

import { CloudInitGenerator } from "../src/utils/cloudInit";
import type { CloudInitForm } from "../src/utils/cloudInit";

describe("CloudInitGenerator", () => {
  const baseFormData = {
    disableSSHPassword: false,
    enableRemoteVideo: true,
    enableSSH: true,
    enableTelemetry: true,
    hostname: "pn-test",
    keyboardLayout: "us",
    password: "testing1234",
    saveSettings: false,
    timezone: "America/Los_Angeles",
    username: "testuser",
    wifiCountry: "US",
    wifiPassword: "testing3456",
    wifiSSID: "test-ssid",
    wifiSSIDHidden: false,
  } as CloudInitForm;
  test("should generate user-data", () => {
    const generator = new CloudInitGenerator(baseFormData);
    const yamlData = generator.generateUserData();

    expect(yamlData).toMatchSnapshot();
  });

  test("should hash password", () => {
    const hashed = CloudInitGenerator.hashPassword(baseFormData.password);
    expect(CloudInitGenerator.comparePassword(baseFormData.password, hashed)).toEqual(true);
  });
});
