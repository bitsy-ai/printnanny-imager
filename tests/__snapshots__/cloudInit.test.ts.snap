// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloudInitGenerator should generate user-data 1`] = `
"#cloud-config
hostname: pn-test
manage_etc_hosts: true
users:
  - default
  - name: testuser
    groups:
      - adm
      - audio
      - cdrom
      - games
      - gpio
      - i2c
      - input
      - netdev
      - plugdev
      - render
      - spi
      - sudo
      - users
      - video
    lock_passwd: false
    passwd: testing1234
    shell: /bin/bash
    sudo: ALL=(ALL) NOPASSWD:ALL
    ssh_authorized_keys: []
timezone: America/Los_Angeles

"
`;
